python/tvm/relay/analysis.py

```diff
@@ -408,3 +411,116 @@ def structural_hash(value):
         msg = ("found value of type {0} expected" +
                "relay.Expr or relay.Type").format(type(value))
         raise TypeError(msg)
+
+def _export_as_relayviz(expr):
+    """Export a Relay function as a nested dictionary, following the RelayViz spec
+    (https://discuss.tvm.ai/t/rfc-visualizing-relay-program-as-graph/4825/10). The dictionary will
+    contain all information useful for visualizing the Relay program and is meant to be consumed
+    by other visualizers.
+
+    Parameters
+    ----------
+    expr : tvm.relay.Expr
+        The input expression.
+
+    Returns
+    -------
+    viz : dict
+        Nested dictionary
+    """
+
+    # node_dict maps a Relay node to an index (node ID)
+    def _traverse_expr(node, node_dict):
+        if node in node_dict:
+            return
+        node_dict[node] = len(node_dict)
+
+    node_dict = {}
+    post_order_visit(expr, lambda x: _traverse_expr(x, node_dict))
+
+    relayviz_nodes = []
+
+    # Sort by node ID
+    for node, node_idx in sorted(node_dict.items(), key=lambda x: x[1]):
+        if isinstance(node, Function):
+            relayviz_nodes.append({
+                'node_kind': 'Function',
+                'body': node_dict[node.body],
+                'params': [node_dict[x] for x in node.params],
+                'ret_type': {
+                    'dtype': node.ret_type.dtype,
+                    'shape': [int(x) for x in node.ret_type.shape]
+                }
```

Comment:
How about there is a type attribute on every op and it contains a string that describes the type of what the node yields? It can be an empty string if there is no value. Any visualizer that wants to actually understand the structure of e.g. types can just deal with Relay directly.

Reply:
@broune I don't see a way to obtain the return type of the op. In fact, it looks like the "ret_type" field in the Function node is optional. It appears that type inference occurs when the Relay program gets compiled.

Reply:
One of the main use cases for this is to print out the state of the IR as it gets mutated through compilation, where type inference is one of the first things (IIRC actually the first thing) that happens, so it is good to show the type if it is present - that accurately reflects the IR.

Reply:
I'll have to ask TVM experts for help. Specifically: 1) where type inference occurs, and 2) how to get the return type of each Op.